File: src/chipschallenge/blockreactions/canMoveNoSlip.java
package chipschallenge.blockreactions;

import chipschallenge.Block;
import chipschallenge.Move.Moves;

/**
 * Standard behavior for when things can move
 */
public class canMoveNoSlip extends BlockReaction {

    private canMoveNoSlip() {
    }
    private static canMoveNoSlip mInstance = null;

    public static synchronized canMoveNoSlip getInstance() {
        if (mInstance == null) {
            mInstance = new canMoveNoSlip();
        }
        return mInstance;
    }

    public void react(Block moving, Block standing) {
        // No reaction
    }

    public final boolean canMove(Block moving, Block standing) {
        return true;
    }

    @Override
    public Moves causesSlip(Block moving, Block standing) {
        return null;
    }
}

File: ChipsChallenge.Shared/Blockreactions/canMoveNoSlip.cs
namespace ChipsChallenge.Shared.Blockreactions
{
    /// <summary>
    /// Standard behavior for when things can move
    /// </summary>
    public class CanMoveNoSlip : NoSlipReaction
    {
        private static CanMoveNoSlip instance;

        private CanMoveNoSlip()
        {
        }

        public static CanMoveNoSlip Instance
        {
            get
            {
                lock (typeof(CanMoveNoSlip))
                {
                    return instance ?? (instance = new CanMoveNoSlip());
                }
            }
        }

        public override void React(Block moving, Block standing)
        {
        }

        public override bool canMove(Block moving, Block standing)
        {
            return true;
        }
    }
}